src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardDictionaryRemoveByContainsKey.cs

```diff
@@ -0,0 +1,120 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Linq;
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+using Resx = Microsoft.NetCore.Analyzers.MicrosoftNetCoreAnalyzersResources;
+
+namespace Microsoft.NetCore.Analyzers.Performance
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class DoNotGuardDictionaryRemoveByContainsKey : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA1839";
+
+        private static readonly LocalizableString s_localizableTitle =
+            new LocalizableResourceString(nameof(Resx.DoNotGuardDictionaryRemoveByContainsKeyTitle), Resx.ResourceManager, typeof(Resx));
+        private static readonly LocalizableString s_localizableMessage =
+            new LocalizableResourceString(nameof(Resx.DoNotGuardDictionaryRemoveByContainsKeyMessage), Resx.ResourceManager, typeof(Resx));
+        private static readonly LocalizableString s_localizableDescription =
+            new LocalizableResourceString(nameof(Resx.DoNotGuardDictionaryRemoveByContainsKeyDescription), Resx.ResourceManager, typeof(Resx));
+
+        public const string AdditionalDocumentLocationInfoSeparator = ";;";
+
+        public const string ConditionalOperation = nameof(ConditionalOperation);
+        public const string ChildStatementOperation = nameof(ChildStatementOperation);
+
+        internal static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            s_localizableDescription,
+            isPortedFxCopRule: false,
+            isDataflowRule: false,
+            additionalCustomTags: WellKnownDiagnosticTags.Unnecessary);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private static void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            var compilation = context.Compilation;
+
+            if (!compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericDictionary2, out var dictionaryType))
+                return;
+
+            context.RegisterOperationAction(AnalyzeOperation, OperationKind.Invocation);
```

Comment:
I think it is better to match conditional operation, then you can use the IConditionalOperation directly instead of checking the parent of all invocation
```suggestion
            context.RegisterOperationAction(AnalyzeOperation, OperationKind.Conditional);
```